src/io/exporters/FBXModel.js

```diff
@@ -0,0 +1,179 @@
+    this.lastIdx = 0;
+  }
+
+  init(vertsCount, indsCount) {
+    this.positions = new Float32Array(vertsCount * FBX_POS_SIZE);
+    this.normals = new Float32Array(vertsCount * FBX_NORM_SIZE);
+    this.colors = new Float32Array(vertsCount * FBX_COL_SIZE);
+    this.indices = new Int32Array(indsCount);
+  }
+
+  setPositions(src, count) {
+    const dst = {
+      array: this.positions,
+      start: this.lastPos,
+      stride: FBX_POS_SIZE,
+    };
+    src.start = 0;
+    setSubArray(src, dst, count, copyFbxPoint3);
+    this.lastPos += count * FBX_POS_SIZE;
+  }
+
+  setTransformedPositions(src, count, matrix) {
+    let idx = this.lastPos;
+    let arrIdx = src.start;
+    for (let i = 0; i < count; ++i, arrIdx += src.stride, idx += FBX_POS_SIZE) {
+      copyTransformedPoint3(src.array, arrIdx, this.positions, idx, { matrix, w: 1 });
+    }
+    this.lastPos += count * FBX_POS_SIZE;
+  }
+
+  setNormals(src, count) {
+    const dst = {
+      array: this.normals,
+      start: this.lastNorm,
+      stride: FBX_NORM_SIZE,
+    };
+    setSubArray(src, dst, count, copyFbxPoint3);
+    this.lastNorm += count * FBX_NORM_SIZE;
+  }
+
+  setTransformedNormals(src, count, matrix) {
+    let idx = this.lastNorm;
+    let arrIdx = src.start;
+    for (let i = 0; i < count; ++i, arrIdx += src.stride, idx += FBX_NORM_SIZE) {
+      copyTransformedPoint3(src.array, arrIdx, this.normals, idx, { matrix, w: 0 });
+    }
+    this.lastNorm += count * FBX_NORM_SIZE;
+  }
+
+  setColors(src, count) {
+    const dst = {
+      array: this.colors,
+      start: this.lastCol,
+      stride: FBX_COL_SIZE,
+    };
+    setSubArray(src, dst, count, copyFbxPoint4, 1);
+    this.lastCol += count * FBX_COL_SIZE;
+  }
+
+  setIndices(array, start, count) {
+    this.indices.set(array, this.lastIdx);
+    this.lastIdx += count;
+  }
+
+  setShiftedIndices(array, count, shift) {
+    const shifted = array.map((x) => x + shift);
+    this.setIndices(shifted, 0, count);
+  }
+
+  getVerticesNumber() {
+    return this.lastPos / FBX_POS_SIZE;
+  }
+
+  addInstance(matrix, geo) {
+    // add indices at first to take old number of vertices for shift
+    const currentCount = this.getVerticesNumber();
+    this.setShiftedIndices(geo.indices, geo.indices.length, currentCount);
+    // simply write vertices at empty space
+    const size = geo.itemSize;
+    this.setTransformedPositions({
+      array: geo.positions,
+      stride: size.position,
+    }, geo.vertsCount, matrix);
+    this.setTransformedNormals({
+      array: geo.normals,
+      stride: size.normal,
+    }, geo.vertsCount, matrix);
+    this.setColors({
+      array: geo.colors,
+      stride: size.color,
+    },
+    geo.vertsCount);
+  }
+
+  concatenate(model) {
+    // store number of vertices before addition
+    const currentCount = this.getVerticesNumber();
+    // add vertices by extending existed arrays
+    this.positions = utils.ConcatTypedArraysUnsafe(this.positions, model.positions);
```

Comment:
How many times do you call this `concatenate()` method? If you call it repeatedly then you face constant memory reallocation and copying. I'd suggest keeping an array of all typed arrays and concatenate them just once in the end (one realloc and one copy).

Reply:
Oh! yes, in the case of memory economy, it's very important. Now, I store an array of subarrays and join them just once.